lib/internal/crypto/keys.js

```diff
@@ -399,6 +403,113 @@ function getKeyTypes(allowKeyObject, bufferOnly = false) {
     ...(!bufferOnly && allowKeyObject ? ['KeyObject', 'CryptoKey'] : [])];
 }
 
+function getKeyObjectHandleFromJwk(key, ctx) {
+  validateObject(key, 'key');
+  validateOneOf(
+    key.kty, 'key.kty', ['RSA', 'EC', 'OKP']);
+
+  const isPublic = ctx === kConsumePublic || ctx === kCreatePublic;
+
+  key = { ...key };
+
+  if (key.kty === 'OKP') {
+    validateString(key.crv, 'key.crv');
+    validateOneOf(
+      key.crv, 'key.crv', ['Ed25519', 'Ed448', 'X25519', 'X448']);
+    validateString(key.x, 'key.x');
+    if (!isPublic) {
+      validateString(key.d, 'key.d');
+    }
+
+    let keyData;
+    if (isPublic)
+      keyData = Buffer.from(key.x, 'base64');
+    else
+      keyData = Buffer.from(key.d, 'base64');
+
+    switch (key.crv) {
+      case 'Ed25519':
+      case 'X25519':
+        if (keyData.byteLength !== 32) {
+          throw new ERR_CRYPTO_INVALID_JWK('Invalid JWK data');
+        }
+        break;
+      case 'Ed448':
+        if (keyData.byteLength !== 57) {
+          throw new ERR_CRYPTO_INVALID_JWK('Invalid JWK data');
+        }
+        break;
+      case 'X448':
+        if (keyData.byteLength !== 56) {
+          throw new ERR_CRYPTO_INVALID_JWK('Invalid JWK data');
+        }
+        break;
+    }
+
+    const handle = new KeyObjectHandle();
+    if (isPublic) {
+      handle.initEDRaw(
+        `NODE-${key.crv.toUpperCase()}`,
+        keyData,
+        kKeyTypePublic);
+    } else {
+      handle.initEDRaw(
+        `NODE-${key.crv.toUpperCase()}`,
+        keyData,
+        kKeyTypePrivate);
+    }
+
+    return handle;
+  }
+
+  if (key.kty === 'EC') {
+    validateString(key.crv, 'key.crv');
+    validateOneOf(
+      key.crv, 'key.crv', ['P-256', 'secp256k1', 'P-384', 'P-521']);
+    validateString(key.x, 'key.x');
+    validateString(key.y, 'key.y');
+
+    if (isPublic) {
+      delete key.d;
+    } else {
+      validateString(key.d, 'key.d');
+    }
+
+    const handle = new KeyObjectHandle();
+    const type = handle.initJwk(key, key.crv);
+    if (type === undefined)
+      throw new ERR_CRYPTO_INVALID_JWK('Invalid JWK data');
+
+    return handle;
+  }
+
+  // RSA
+  validateString(key.n, 'key.n');
+  validateString(key.e, 'key.e');
+  if (isPublic) {
+    delete key.d;
+    delete key.p;
+    delete key.q;
+    delete key.dp;
+    delete key.dq;
+    delete key.qi;
+  } else {
+    validateString(key.d, 'key.d');
+    validateString(key.p, 'key.p');
+    validateString(key.q, 'key.q');
+    validateString(key.dp, 'key.dp');
+    validateString(key.dq, 'key.dq');
+    validateString(key.qi, 'key.qi');
+  }
+
+  const handle = new KeyObjectHandle();
```

Comment:
Not sure how I feel about creating `KeyObjectHandle`s in more places.

Reply:
Do you have a concrete suggestion? I'm only using what's available.

Reply:
Not right now, sorry. We can probably improve that later. I don't think this internal inconsistency has any visible effect to the user, but I'm not entirely sure.